File: Client/main/java/com/example/securechat/crypto/ECDSASignature.java
package com.example.securechat.crypto;

import android.util.Log;
import org.json.JSONObject;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;

public class ECDSASignature extends SignatureBase {
    private static final String TAG = "ECDSASignature";

    // P-192 parameters (same as ECDH)
    private static final BigInteger CURVE_P = new BigInteger("6277101735386680763835789423207666416083908700390324961279");
    private static final BigInteger CURVE_A = BigInteger.valueOf(-3);
    private static final BigInteger CURVE_B = new BigInteger("2455155546008943817740293915197451784769108058161191238065");
    private static final BigInteger CURVE_GX = new BigInteger("3289624317623424368845348028842487418520868978772050262753");
    private static final BigInteger CURVE_GY = new BigInteger("5673242899673324591834582889556471730778853907191064256384");
    private static final BigInteger CURVE_ORDER = new BigInteger("6277101735386680763835789423176059013767194773182842284081");

    public ECDSASignature() {
        super("ECDSA-P192");
    }

    public static class ECPoint {
        public BigInteger x, y;

        public ECPoint(BigInteger x, BigInteger y) {
            this.x = x;
            this.y = y;
        }
    }

    // Concrete implementation of Signature for ECDSA
    public static class ECDSASignatureData extends SignatureBase.Signature {
        public String r, s, messageHash;

        public ECDSASignatureData(String r, String s, String messageHash) {
            super("ECDSA-P192");
            this.r = r;
            this.s = s;
            this.messageHash = messageHash;
        }

        @Override
        public JSONObject toJSON() throws Exception {
            JSONObject json = new JSONObject();
            json.put("r", r);
            json.put("s", s);
            json.put("messageHash", messageHash);
            json.put("algorithm", algorithm);
            return json;
        }

        public static ECDSASignatureData fromJSON(JSONObject json) throws Exception {
            return new ECDSASignatureData(
                    json.getString("r"),
                    json.getString("s"),
                    json.getString("messageHash")
            );
        }
    }

    @Override
    public KeyPair generateSignatureKeyPair() {
        SecureRandom random = new SecureRandom();
        byte[] privateKeyBytes = new byte[24]; // 192 bits
        random.nextBytes(privateKeyBytes);

        BigInteger privateKey = new BigInteger(1, privateKeyBytes)
                .mod(CURVE_ORDER.subtract(BigInteger.ONE))
                .add(BigInteger.ONE);

        ECPoint publicKeyPoint = scalarMult(new ECPoint(CURVE_GX, CURVE_GY), privateKey);

        try {
            JSONObject publicKey = new JSONObject();
            publicKey.put("x", publicKeyPoint.x.toString());
            publicKey.put("y", publicKeyPoint.y.toString());

            return new KeyPair(privateKey.toString(), publicKey);
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate ECDSA signature key pair", e);
        }
    }

    @Override
    public Signature signMessage(String message, String privateKeyStr) throws Exception {
        BigInteger messageHash = hashMessage(message);
        BigInteger privateKey = new BigInteger(privateKeyStr);

        // Generate random k
        SecureRandom random = new SecureRandom();
        BigInteger k;
        do {
            byte[] kBytes = new byte[24];
            random.nextBytes(kBytes);
            k = new BigInteger(1, kBytes).mod(CURVE_ORDER.subtract(BigInteger.ONE)).add(BigInteger.ONE);
        } while (k.equals(BigInteger.ZERO));

        // Calculate r = (k * G).x mod order
        ECPoint kG = scalarMult(new ECPoint(CURVE_GX, CURVE_GY), k);
        BigInteger r = kG.x.mod(CURVE_ORDER);

        if (r.equals(BigInteger.ZERO)) {
            throw new Exception("Invalid ECDSA signature generation (r = 0)");
        }

        // Calculate s = k^(-1) * (hash + r * privateKey) mod order
        BigInteger kInv = k.modInverse(CURVE_ORDER);
        BigInteger s = kInv.multiply(messageHash.add(r.multiply(privateKey))).mod(CURVE_ORDER);

        if (s.equals(BigInteger.ZERO)) {
            throw new Exception("Invalid ECDSA signature generation (s = 0)");
        }

        return new ECDSASignatureData(r.toString(), s.toString(), messageHash.toString());
    }

    @Override
    public boolean verifySignature(String message, Signature signature, JSONObject publicKeyJson) {
        try {
            // Cast to ECDSASignatureData
            if (!(signature instanceof ECDSASignatureData)) {
                Log.e(TAG, "Invalid signature type for ECDSA verification");
                return false;
            }

            ECDSASignatureData ecdsaSig = (ECDSASignatureData) signature;
            BigInteger messageHash = hashMessage(message);
            BigInteger r = new BigInteger(ecdsaSig.r);
            BigInteger s = new BigInteger(ecdsaSig.s);

            BigInteger pubX = new BigInteger(publicKeyJson.getString("x"));
            BigInteger pubY = new BigInteger(publicKeyJson.getString("y"));
            ECPoint publicKey = new ECPoint(pubX, pubY);

            // Verify r and s are in valid range
            if (r.compareTo(BigInteger.ZERO) <= 0 || r.compareTo(CURVE_ORDER) >= 0 ||
                    s.compareTo(BigInteger.ZERO) <= 0 || s.compareTo(CURVE_ORDER) >= 0) {
                return false;
            }

            // Calculate w = s^(-1) mod order
            BigInteger w = s.modInverse(CURVE_ORDER);

            // Calculate u1 = hash * w mod order
            BigInteger u1 = messageHash.multiply(w).mod(CURVE_ORDER);

            // Calculate u2 = r * w mod order
            BigInteger u2 = r.multiply(w).mod(CURVE_ORDER);

            // Calculate point = u1 * G + u2 * publicKey
            ECPoint point1 = scalarMult(new ECPoint(CURVE_GX, CURVE_GY), u1);
            ECPoint point2 = scalarMult(publicKey, u2);
            ECPoint point = pointAdd(point1, point2);

            if (point == null) return false;

            // Verify r == point.x mod order
            BigInteger v = point.x.mod(CURVE_ORDER);
            return v.equals(r);

        } catch (Exception e) {
            Log.e(TAG, "ECDSA signature verification error", e);
            return false;
        }
    }

    @Override
    public boolean validateSignature(Signature signature) {
        if (!(signature instanceof ECDSASignatureData)) {
            return false;
        }
        ECDSASignatureData ecdsaSig = (ECDSASignatureData) signature;
        return ecdsaSig.r != null && ecdsaSig.s != null &&
                algorithmName.equals(ecdsaSig.algorithm);
    }

    // Hash message using SHA-256
    public BigInteger hashMessage(String message) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        byte[] hashBytes = digest.digest(message.getBytes(StandardCharsets.UTF_8));

        // Convert to BigInteger
        BigInteger hash = new BigInteger(1, hashBytes);

        // Reduce to curve order
        return hash.mod(CURVE_ORDER);
    }

    @Override
    public int getSignatureSize() {
        return 384; // 192 * 2 bits for r and s
    }

    @Override
    public JSONObject getParameters() throws Exception {
        JSONObject params = new JSONObject();
        params.put("algorithm", algorithmName);
        params.put("curve", "P-192");
        params.put("signatureSize", getSignatureSize());
        params.put("hashAlgorithm", "SHA-256");
        params.put("p", CURVE_P.toString());
        params.put("a", CURVE_A.toString());
        params.put("b", CURVE_B.toString());
        params.put("gx", CURVE_GX.toString());
        params.put("gy", CURVE_GY.toString());
        params.put("order", CURVE_ORDER.toString());
        return params;
    }

    // Helper methods (same as ECDH implementation)
    private ECPoint pointAdd(ECPoint pointP, ECPoint pointQ) {
        if (pointP == null) return pointQ;
        if (pointQ == null) return pointP;

        if (pointP.x.equals(pointQ.x)) {
            if (pointP.y.add(pointQ.y).mod(CURVE_P).equals(BigInteger.ZERO)) {
                return null;
            }
            if (pointP.y.equals(pointQ.y)) {
                return pointDouble(pointP);
            }
        }

        BigInteger deltaY = pointQ.y.subtract(pointP.y);
        BigInteger deltaX = pointQ.x.subtract(pointP.x);
        BigInteger slope = deltaY.multiply(deltaX.modInverse(CURVE_P)).mod(CURVE_P);

        BigInteger resultX = slope.multiply(slope).subtract(pointP.x).subtract(pointQ.x).mod(CURVE_P);
        BigInteger resultY = slope.multiply(pointP.x.subtract(resultX)).subtract(pointP.y).mod(CURVE_P);

        if (resultX.compareTo(BigInteger.ZERO) < 0) {
            resultX = resultX.add(CURVE_P);
        }
        if (resultY.compareTo(BigInteger.ZERO) < 0) {
            resultY = resultY.add(CURVE_P);
        }

        return new ECPoint(resultX, resultY);
    }

    private ECPoint pointDouble(ECPoint point) {
        if (point == null) return null;

        BigInteger numerator = point.x.multiply(point.x).multiply(BigInteger.valueOf(3)).add(CURVE_A);
        BigInteger denominator = point.y.multiply(BigInteger.valueOf(2));
        BigInteger slope = numerator.multiply(denominator.modInverse(CURVE_P)).mod(CURVE_P);

        BigInteger resultX = slope.multiply(slope).subtract(point.x.multiply(BigInteger.valueOf(2))).mod(CURVE_P);
        BigInteger resultY = slope.multiply(point.x.subtract(resultX)).subtract(point.y).mod(CURVE_P);

        if (resultX.compareTo(BigInteger.ZERO) < 0) {
            resultX = resultX.add(CURVE_P);
        }
        if (resultY.compareTo(BigInteger.ZERO) < 0) {
            resultY = resultY.add(CURVE_P);
        }

        return new ECPoint(resultX, resultY);
    }

    private ECPoint scalarMult(ECPoint point, BigInteger scalar) {
        if (point == null || scalar.equals(BigInteger.ZERO)) {
            return null;
        }

        if (scalar.equals(BigInteger.ONE)) {
            return new ECPoint(point.x, point.y);
        }

        ECPoint result = null;
        ECPoint addend = new ECPoint(point.x, point.y);
        BigInteger k = scalar;

        while (k.compareTo(BigInteger.ZERO) > 0) {
            if (k.testBit(0)) {
                result = pointAdd(result, addend);
            }
            addend = pointDouble(addend);
            k = k.shiftRight(1);
        }

        return result;
    }
}

File: Server/src/crypto/ecdsaSignature.js
// ============================================
// File: src/crypto/ecdsaSignature.js
// ============================================

import { SignatureBase } from './signatureBase.js';
import { SecureRandom } from './secureRandom.js';

export class ECDSASignature extends SignatureBase {
  constructor(curveParams = null) {
    super("ECDSA-P192");
    
    // Use curve parameters from ECDH or defaults
    if (curveParams) {
      this.p = BigInt(curveParams.p);
      this.a = BigInt(curveParams.a);
      this.b = BigInt(curveParams.b);
      this.Gx = BigInt(curveParams.Gx);
      this.Gy = BigInt(curveParams.Gy);
      this.order = BigInt(curveParams.order);
    } else {
      // P-192 default parameters
      this.p = 6277101735386680763835789423207666416083908700390324961279n;
      this.a = -3n;
      this.b = 2455155546008943817740293915197451784769108058161191238065n;
      this.Gx = 3289624317623424368845348028842487418520868978772050262753n;
      this.Gy = 5673242899673324591834582889556471730778853907191064256384n;
      this.order = 6277101735386680763835789423176059013767194773182842284081n;
    }
    
    console.log('ECDSA Signature initialized for P-192 curve');
  }

  // Generate signature key pair (different from ECDH keys)
  generateSignatureKeyPair() {
    // Use deterministic key for demo, crypto.getRandomValues for production
    const privateKey = this.generatePrivateKey();
    const publicKey = this.generatePublicKey(privateKey);
    
    return {
      privateKey: privateKey.toString(),
      publicKey: this.serializePublicKey(publicKey)
    };
  }

  generatePrivateKey() {
    return SecureRandom.generatePrivateKey(this.order);
  }

  generatePublicKey(privateKey) {
    const G = { x: this.Gx, y: this.Gy };
    return this.scalarMult(G, privateKey);
  }

  // Hash message using SHA-256
  async hashMessage(message) {
    const encoder = new TextEncoder();
    const data = encoder.encode(message);
    const hashBuffer = await crypto.subtle.digest('SHA-256', data);
    const hashArray = new Uint8Array(hashBuffer);
    
    // Convert to BigInt
    let hash = 0n;
    for (let i = 0; i < hashArray.length; i++) {
      hash = (hash << 8n) + BigInt(hashArray[i]);
    }
    
    return hash % this.order;
  }

  // Sign message with ECDSA
  async signMessage(message, privateKey) {
    const messageHash = await this.hashMessage(message);
    const privKey = BigInt(privateKey);
    
    // Generate deterministic k for demo
    let k = await SecureRandom.generateSecureK(messageHash, privKey, this.order);
    
    // Calculate r = (k * G).x mod order
    const G = { x: this.Gx, y: this.Gy };
    const kG = this.scalarMult(G, k);
    const r = kG.x % this.order;
    
    if (r === 0n) {
      throw new Error('Invalid signature generation (r = 0)');
    }
    
    // Calculate s = k^(-1) * (hash + r * privateKey) mod order
    const kInv = this.modInverse(k, this.order);
    const s = (kInv * (messageHash + r * privKey)) % this.order;
    
    if (s === 0n) {
      throw new Error('Invalid signature generation (s = 0)');
    }
    
    return {
      r: r.toString(),
      s: s.toString(),
      messageHash: messageHash.toString(),
      algorithm: this.algorithmName
    };
  }

  // Verify ECDSA signature
  async verifySignature(message, signature, publicKey) {
    try {
      const messageHash = await this.hashMessage(message);
      const r = BigInt(signature.r);
      const s = BigInt(signature.s);
      const pubKey = this.deserializePublicKey(publicKey);
      
      // Verify r and s are in valid range
      if (r <= 0n || r >= this.order || s <= 0n || s >= this.order) {
        return false;
      }
      
      // Calculate w = s^(-1) mod order
      const w = this.modInverse(s, this.order);
      
      // Calculate u1 = hash * w mod order
      const u1 = (messageHash * w) % this.order;
      
      // Calculate u2 = r * w mod order
      const u2 = (r * w) % this.order;
      
      // Calculate point = u1 * G + u2 * publicKey
      const G = { x: this.Gx, y: this.Gy };
      const point1 = this.scalarMult(G, u1);
      const point2 = this.scalarMult(pubKey, u2);
      const point = this.pointAdd(point1, point2);
      
      if (!point) return false;
      
      // Verify r == point.x mod order
      const v = point.x % this.order;
      return v === r;
      
    } catch (error) {
      console.error('ECDSA signature verification error:', error);
      return false;
    }
  }

  serializeSignature(signature) {
    return {
      r: signature.r,
      s: signature.s,
      messageHash: signature.messageHash,
      algorithm: signature.algorithm
    };
  }

  deserializeSignature(data) {
    return {
      r: data.r,
      s: data.s,
      messageHash: data.messageHash,
      algorithm: data.algorithm
    };
  }

  serializePublicKey(publicKey) {
    return {
      x: publicKey.x.toString(),
      y: publicKey.y.toString()
    };
  }

  deserializePublicKey(data) {
    return {
      x: BigInt(data.x),
      y: BigInt(data.y)
    };
  }

  validateSignature(signature) {
    return signature.r && signature.s && signature.algorithm === this.algorithmName;
  }

  getSignatureSize() {
    return 384; // 192 * 2 bits for r and s
  }

  getParameters() {
    return {
      algorithm: this.algorithmName,
      curve: "P-192",
      signatureSize: this.getSignatureSize(),
      hashAlgorithm: "SHA-256",
      p: this.p.toString(),
      a: this.a.toString(),
      b: this.b.toString(),
      G: {
        x: this.Gx.toString(),
        y: this.Gy.toString()
      },
      order: this.order.toString()
    };
  }

  // Helper methods (same as ECDH implementation)
  modInverse(a, m) {
    a = (a % m + m) % m;
    let [old_r, r] = [a, m];
    let [old_s, s] = [1n, 0n];

    while (r !== 0n) {
      const quotient = old_r / r;
      [old_r, r] = [r, old_r - quotient * r];
      [old_s, s] = [s, old_s - quotient * s];
    }

    return (old_s % m + m) % m;
  }

  pointAdd(P, Q) {
    if (!P) return Q;
    if (!Q) return P;

    if (P.x === Q.x && (P.y + Q.y) % this.p === 0n) {
      return null;
    }

    let lambda;
    if (P.x === Q.x && P.y === Q.y) {
      lambda = (3n * P.x * P.x + this.a) * this.modInverse(2n * P.y, this.p) % this.p;
    } else {
      lambda = (Q.y - P.y) * this.modInverse((Q.x - P.x + this.p) % this.p, this.p) % this.p;
    }

    const Rx = (lambda * lambda - P.x - Q.x) % this.p;
    const Ry = (lambda * (P.x - Rx) - P.y) % this.p;

    return {
      x: (Rx % this.p + this.p) % this.p,
      y: (Ry % this.p + this.p) % this.p
    };
  }

  scalarMult(P, n) {
    let Q = P;
    let R = null;
    let k = n;

    while (k > 0n) {
      if (k % 2n === 1n) {
        R = this.pointAdd(R, Q);
      }
      Q = this.pointAdd(Q, Q);
      k = k / 2n;
    }

    return R;
  }
}